File: src/java/MatrixMult3.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;

class MatrixMult3
{
	/**
	 * container/parser for arguments 
	 */
	public static class Configuration
	{
		public final File file;
		public final int dimension;
		public final boolean dryRun;
		public final boolean asFloat;
		
		public Configuration(String[] args)
		{
			if(args.length < 2 || args[0].length() == 0 || args[1].length() == 0)
			{
				throw new IllegalArgumentException("File name must be provided in the first argument, and dimension in the second");
			}
			
			file = new File(args[0]);
			dimension = Integer.parseInt(args[1]);
			
			boolean dryRun = false;
			boolean asFloat = false;
			
			for(int i = 2; i < args.length; i++)
			{
				if("--dry-run".equals(args[i]))
				{
					if(dryRun)
					{
						throw new IllegalArgumentException("Argument \"--dry-run\" provided multiple times.");
					}
					else
					{
						dryRun = true;
					}
				}
				else if("--float".equals(args[i]))
				{
					if(asFloat)
					{
						throw new IllegalArgumentException("Argument \"--float\" provided multiple times.");
					}
					else
					{
						asFloat = true;
					}
				}
				else
				{
					throw new IllegalArgumentException("Unknown argument: \"" + args[i] + "\".");
				}
			}
			
			this.dryRun = dryRun;
			this.asFloat = asFloat;
		}
	}
	
	public static class NumberSource
	{
		private final BufferedReader bufferedReader;
		
		boolean isEmpty;
		
		public NumberSource(File file) throws IOException
		{
			isEmpty = false;
			
			Charset us_ascii = Charset.forName("US-ASCII");
			
			// read file entirely into memory
			
			bufferedReader = new BufferedReader(new InputStreamReader(new FileInputStream(file), us_ascii));
			seekToNonWhite();
		}
		
		public void skipRestOfLine() throws IOException
		{
			int value;
			while(true)
			{
				value = bufferedReader.read();
				
				if(value < 0)
				{
					isEmpty = true;
					return;
				}
				else if(((char)value) == '\n')
				{
					seekToNonWhite();
					return;
				}
			}
		}
		
		private void seekToNonWhite() throws IOException
		{
			int value;
			while(true)
			{
				bufferedReader.mark(1);
				value = bufferedReader.read();
				
				if(value < 0)
				{
					isEmpty = true;
					return;
				}
				else if(!Character.isWhitespace((char)value) || ((char)value) == '\n')
				{
					bufferedReader.reset();
					return;
				}
			}
		}
		
		private String getNumber() throws IOException
		{
			if(isEmpty)
			{
				throw new IllegalArgumentException("Provided input file is too short.");
			}
			
			int value;
			StringBuilder stringBuilder = new StringBuilder();
			while(true)
			{
				bufferedReader.mark(1);
				value = bufferedReader.read();
				
				if(((char)value) == '\n')
				{
					throw new IllegalArgumentException("Unexpected line break.");
				}
				else if(value < 0)
				{
					isEmpty = true;
					break;
				}
				else if(Character.isWhitespace((char)value))
				{
					seekToNonWhite();
					break;
				}
				else
				{
					stringBuilder.append((char)value);
				}
			}
			
			return stringBuilder.toString();
		}
		
		public int getInt() throws NumberFormatException, IOException
		{
			return Integer.parseInt(getNumber().split("\\.")[0]);
		}
		
		public float getFloat() throws NumberFormatException, IOException
		{
			return Float.parseFloat(getNumber());
		}
		
		public void close() throws IOException
		{
			bufferedReader.close();
		}
	}
	
	public static class IntegerMatrix
	{
		private final int[][] data;
		
		public IntegerMatrix(int dimension)
		{
			data = new int[dimension][dimension];
		}
		
		public IntegerMatrix(int dimension, NumberSource numberSource) throws IOException
		{
			this(dimension);
			
			for(int row_index = 0; row_index < getDimension(); row_index++)
			{
				for(int cell_index = 0; cell_index < getDimension(); cell_index++)
				{
					data[row_index][cell_index] = numberSource.getInt();
				}
				numberSource.skipRestOfLine();
			}
		}
		
		public int getDimension()
		{
			return data.length;
		}
		
		public static IntegerMatrix multiply(IntegerMatrix a, IntegerMatrix b)
		{
			if(a.getDimension() != b.getDimension())
			{
				throw new IllegalArgumentException("Matrices must be of the same dimension");
			}
			
			IntegerMatrix ret = new IntegerMatrix(a.getDimension());
			
			for(int row_index = 0; row_index < ret.getDimension(); row_index++)
			{
				for(int cell_index = 0; cell_index < ret.getDimension(); cell_index++)
				{
					for(int k = 0; k < ret.getDimension(); k++)
					{
						ret.data[row_index][cell_index] += a.data[row_index][k] * b.data[k][cell_index];
					}
				}
			}
			
			return ret;
		}
	}
	
	public static class FloatMatrix
	{
		private final float[][] data;
		
		public FloatMatrix(int dimension)
		{
			data = new float[dimension][dimension];
		}
		
		public FloatMatrix(int dimension, NumberSource numberSource) throws IOException
		{
			this(dimension);
			
			for(int row_index = 0; row_index < getDimension(); row_index++)
			{
				for(int cell_index = 0; cell_index < getDimension(); cell_index++)
				{
					data[row_index][cell_index] = numberSource.getFloat();
				}
				numberSource.skipRestOfLine();
			}
		}
		
		public int getDimension()
		{
			return data.length;
		}
		
		public static FloatMatrix multiply(FloatMatrix a, FloatMatrix b)
		{
			if(a.getDimension() != b.getDimension())
			{
				throw new IllegalArgumentException("Matrices must be of the same dimension");
			}
			
			FloatMatrix ret = new FloatMatrix(a.getDimension());
			
			for(int row_index = 0; row_index < ret.getDimension(); row_index++)
			{
				for(int cell_index = 0; cell_index < ret.getDimension(); cell_index++)
				{
					for(int k = 0; k < ret.getDimension(); k++)
					{
						ret.data[row_index][cell_index] += a.data[row_index][k] * b.data[k][cell_index];
					}
				}
			}
			
			return ret;
		}
	}
	
	public static void main(String[] args) throws IOException
	{
		// get configuration
		Configuration configuration = new Configuration(args);
		
		NumberSource numberSource = null;
		FloatMatrix floatMatrix = null;
		IntegerMatrix integerMatrix = null;
		try
		{
			// get number source
			numberSource = new NumberSource(configuration.file);
			
			if(configuration.asFloat)
			{
				floatMatrix = new FloatMatrix(configuration.dimension, numberSource);
			}
			else
			{
				integerMatrix = new IntegerMatrix(configuration.dimension, numberSource);
			}
		}
		finally
		{
			if(numberSource != null)
			{
				numberSource.close();
				numberSource = null;
			}
		}
		
		if(!configuration.dryRun)
		{
			if(configuration.asFloat)
			{
				@SuppressWarnings("unused")
				FloatMatrix floatMatrixResult = FloatMatrix.multiply(floatMatrix, floatMatrix);
			}
			else
			{
				@SuppressWarnings("unused")
				IntegerMatrix integerMatrixResult = IntegerMatrix.multiply(integerMatrix, integerMatrix);
			}
		}
	}
}

File: src/cpp/MatrixMult3.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <stdexcept>

using std::string;
using std::stringstream;
using std::cout;
using std::endl;
using std::runtime_error;
using std::ifstream;

class Configuration
{
	public:
	
	string* filename;
	int dimension;
	bool dryRun;
	bool asFloat;
	
	Configuration(int argc, char** argv)
	{
		string args[argc - 1];
		
		for(int i = 1; i < argc; i++)
		{
			args[i - 1] = string(argv[i]);
		}
		
		if(argc < 2 || args[0].length() == 0 || args[1].length() == 0)
		{
			throw runtime_error("File name must be provided in the first argument, and dimension in the second");
		}
		
		this->filename = new string(args[0]);
		
		stringstream dimension_stringstream(argv[2]);
		dimension_stringstream >> this->dimension;
		if(dimension_stringstream.fail())
		{
			throw runtime_error("dimension not an integer");
		}
		bool dryRun = false;
		bool asFloat = false;
		
		for(int i = 2; i < argc - 1; i++)
		{
			if(args[i].compare("--dry-run") == 0)
			{
				if(dryRun)
				{
					throw runtime_error("Argument \"--dry-run\" provided multiple times.");
				}
				else
				{
					dryRun = true;
				}
			}
			else if(args[i].compare("--float") == 0)
			{
				if(asFloat)
				{
					throw runtime_error("Argument \"--float\" provided multiple times.");
				}
				else
				{
					asFloat = true;
				}
			}
			else
			{
				throw runtime_error("Unknown argument: \"" + args[i] + "\".");
			}
		}
		
		this->dryRun = dryRun;
		this->asFloat = asFloat;
	}
	
	~Configuration()
	{
		delete this->filename;
	}
};

class NumberSource
{
	private:
	
	ifstream* inputFile;
	
	public:
	
	NumberSource(string filename)
	{
		this->inputFile = new ifstream(filename.c_str());
	}
	
	~NumberSource()
	{
		this->inputFile->close();
		delete this->inputFile;
	}
	
	void skipRestOfLine()
	{
		if(this->inputFile->is_open())
		{
			while(this->inputFile->good())
			{
				char c = this->inputFile->get();
				
				if(c == '\n')
				{
					return;
				}
			}
		}
		
		throw runtime_error("can't read number from input file");
	}
	
	int getInt()
	{
		return (int)this->getFloat();
	}
	
	float getFloat()
	{
		float number;
		
		if(this->inputFile->is_open() && this->inputFile->good())
		{
			*this->inputFile >> number;
			if(this->inputFile->fail())
			{
				throw runtime_error("bad number from input file");
			}
			
			return number;
		}
		
		throw runtime_error("can't read number from input file");
	}
};

template<typename T>
class Matrix
{
	private:
	T* data;
	int dimension;
	
	public:
	
	Matrix(int dimension, NumberSource* numberSource = NULL) : dimension(dimension)
	{
		this->data = new T[dimension * dimension];
		
		if(numberSource != NULL)
		{
			for(int row_index = 0; row_index < getDimension(); row_index++)
			{
				for(int cell_index = 0; cell_index < getDimension(); cell_index++)
				{
					this->data[row_index * getDimension() + cell_index] = (T)numberSource->getFloat();
					
					//cout << this->data[row_index * getDimension() + cell_index] << endl;
				}
				
				numberSource->skipRestOfLine();
			}
		}
	}
	
	~Matrix()
	{
		delete this->data;
	}
	
	int getDimension()
	{
		return this->dimension;
	}
	
	static Matrix* multiply(Matrix& a, Matrix& b)
	{
		if(a.getDimension() != b.getDimension())
		{
			throw runtime_error("Matrices must be of the same dimension");
		}
		
		Matrix* ret = new Matrix(a.getDimension());
		
		for(int row_index = 0; row_index < ret->getDimension(); row_index++)
		{
			for(int cell_index = 0; cell_index < ret->getDimension(); cell_index++)
			{
				ret->data[row_index * ret->getDimension() + cell_index] = 0;
				for(int k = 0; k < ret->getDimension(); k++)
				{
					ret->data[row_index * ret->getDimension() + cell_index] += a.data[row_index * ret->getDimension() + k] * b.data[k * ret->getDimension() + cell_index];
				}
			}
		}
		
		return ret;
	}
	
	void print()
	{
		for(int row_index = 0; row_index < getDimension(); row_index++)
		{
			for(int cell_index = 0; cell_index < getDimension(); cell_index++)
			{
				cout << this->data[row_index * getDimension() + cell_index] << "       ";
			}
			
			cout << endl;
		}
	}
};

int main(int argc, char** argv)
{
	Configuration configuration = Configuration(argc, argv);
	
	NumberSource numberSource = NumberSource(*configuration.filename);
	
	if(configuration.asFloat)
	{
		Matrix<float> input(configuration.dimension, &numberSource);
		
		if(!configuration.dryRun)
		{
			Matrix<float>* result = Matrix<float>::multiply(input, input);
			
			//result->print();
			
			delete result;
		}
	}
	else
	{
		Matrix<int> input(configuration.dimension, &numberSource);
		
		if(!configuration.dryRun)
		{
			Matrix<int>* result = Matrix<int>::multiply(input, input);
			//result->print();
			
			delete result;
		}
	}
	
	//cout << "done" << endl;
	return 0;
}
